### PR TITLE
New commandline sample with schema and stable script concept on release/4.0.0

### PR DESCRIPTION
**Pull request by dennisroche**

### CODE REVIEW

File: src/Samples/CommandLineApplication/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# EditorConfig (http://editorconfig.org) is awesome. It defines and maintains consistent coding styles between different editors and IDEs.
+# IDE Extensions
+#   Visual Studio, install https://visualstudiogallery.msdn.microsoft.com/c8bccfe2-650c-4b42-bc5c-845e21f96328
+#   VS Code, install https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig
+root = false
```
Comment (JakeGinnivan) [inline src/Samples/CommandLineApplication/.editorconfig]: Should this be moved to the root of the project?

File: src/Samples/CommandLineApplication/Program.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace CommandLineApplication
+{
+    public class Program
+    {
+        public static int Main(string[] args)
+        {
+            var app = new Microsoft.Extensions.CommandLineUtils.CommandLineApplication
+            {
+                Name = "CommunicareSchemaMigrations.exe",
```
Comment (JakeGinnivan) [inline src/Samples/CommandLineApplication/Program.cs]: Deliberate naming?

File: src/Samples/CommandLineApplication/DbUp/UpgradeEngineAssertions.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            var executedScripts = upgrader.GetExecutedScripts().Select(x => allScripts[x]).ToArray();
+            var newScripts = upgrader.GetScriptsToExecute().ToArray();
+
+            foreach (var executedScript in executedScripts)
+            {
+                var executedScriptHash = CreateHashForScript(executedScript);
+                foreach (var newScript in newScripts)
+                {
+                    var newScriptHash = CreateHashForScript(newScript);
+                    if (executedScriptHash.SequenceEqual(newScriptHash))
```
Comment (JakeGinnivan) [inline src/Samples/CommandLineApplication/DbUp/UpgradeEngineAssertions.cs]: Could this MD5 stuff be moved into Core? With a new extension `.WithoutScriptVerification()` to turn off?

File: src/Samples/CommandLineApplication/SqlServerMigrator.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        public void Create()
+        {
+            EnsureDatabase.For.SqlDatabase(_connectionString, _upgradeLogger);
+        }
+
+        public bool IsUpgradeRequired()
+        {
+            var upgrader = DeployChanges.To
+                .SqlDatabase(_connectionString)
+                .WithScriptsEmbeddedInAssembly(typeof(SqlServerMigrator).Assembly, _scriptNamespace, _stableScriptsOnly)
```
Comment (JakeGinnivan) [inline src/Samples/CommandLineApplication/SqlServerMigrator.cs]: I was thinking about putting a feature into DbUp which ensured scripts are executed in order. How does this scenario work:

Script001.Stable.sql
Script002.sql
Script003.Stable.sql

The order executed is not Script001, Script003, then Script002. If I then want to promote Script002 to stable. The order then becomes Script001, Script002, Script003. The ordering of execution can change based on promoting a script which could be hard to follow?

Moving this feature to be built on https://github.com/DbUp/DbUp/pull/217 could be good, you could mark a version as stable and then we could enforce better semantics?

File: src/Samples/CommandLineApplication/Program.cs (hunk 2 above)
Comment (dennisroche) [inline src/Samples/CommandLineApplication/Program.cs]: opps. delete.

File: src/Samples/CommandLineApplication/.editorconfig (hunk 1 above)
Comment (dennisroche) [inline src/Samples/CommandLineApplication/.editorconfig]: I think we should have an `.editorconfig` in the root of the project. I've recently started having one in the root then one as needed closer to project

File: src/Samples/CommandLineApplication/DbUp/UpgradeEngineAssertions.cs (hunk 3 above)
Comment (dennisroche) [inline src/Samples/CommandLineApplication/DbUp/UpgradeEngineAssertions.cs]: Yeah, it would be a nice feature. I'll pull it out and add an issue to describe the problem I was trying to solve.

File: src/Samples/CommandLineApplication/SqlServerMigrator.cs (hunk 4 above)
Comment (dennisroche) [inline src/Samples/CommandLineApplication/SqlServerMigrator.cs]: I have two assumptions for our stable scripts,

1. Scripts are promoted to stable in order
2. We drop the database before deploying a stable to that environment.

I have our deployment setup in Octopus Deploy that packages with a pre-release semver will drop the database. Once the semver is a stable, it will not drop and only deploy stable scripts. I then use lifecycles to ensure a pre-release cannot be promoted beyond Development environment.

**Comment by droyad** [comment]
Oops, this closed when I deleted the `4.0.0` branch after merging it into `master`

**Comment by dennisroche** [comment]
Merge conflict resolved. 👍

**Comment by droyad** [comment]
I assume this is a how-to-use sample? Should we put it into another repo?

**Comment by droyad** [comment]
Ah nm, I saw the other samples
